model/labels/regexp.go

```diff
@@ -169,38 +176,48 @@ func findSetMatchesInternal(re *syntax.Regexp, base string) (matches []string, c
 		// In some case like negation [^0-9] a lot of possibilities exists and that
 		// can create thousands of possible matches at which points we're better off using regexp.
 		if totalSet > maxSetMatches {
-			return nil, false
+			return nil, nil, false
 		}
 		for i := 0; i+1 < len(re.Rune); i += 2 {
 			lo, hi := re.Rune[i], re.Rune[i+1]
 			for c := lo; c <= hi; c++ {
 				matches = append(matches, base+string(c))
 			}
 		}
-		return matches, isCaseSensitive(re)
+		return matches, nil, isCaseSensitive(re)
 	default:
-		return nil, false
+		return nil, nil, false
 	}
-	return nil, false
+	return nil, nil, false
 }
 
-func findSetMatchesFromConcat(re *syntax.Regexp, base string) (matches []string, matchesCaseSensitive bool) {
+func findSetMatchesFromConcat(re *syntax.Regexp, base string) (matches []string, prefixes []prefixMatcher, matchesCaseSensitive bool) {
 	if len(re.Sub) == 0 {
-		return nil, false
+		return nil, nil, false
 	}
 	clearCapture(re.Sub...)
 
 	matches = []string{base}
 
 	for i := 0; i < len(re.Sub); i++ {
-		var newMatches []string
+		// An OpStar makes the matches so far into prefixes.
+		if i > 0 && re.Sub[i].Op == syntax.OpStar && len(matches) > 0 {
+			prefixes = make([]prefixMatcher, 0, len(matches))
+			for _, prefix := range matches {
+				right := stringMatcherFromRegexpInternal(re.Sub[i])
```

Comment:
Sharing here a feedback I gave privately.

Here we're matching re.Sub[i] as the right side, but what if i is not the last sub of regexp and there are others?

For example this test fails:
```
	const pattern = "a.*a|b.*b|c.*c|d.*d|e.*e|f.*f|g.*g|h.*h|i.*i|l.*l|m.*m|n.*n|o.*o|p.*p|q.*q|r.*r|s.*s|t.*t|u.*u|v.*v|z.*z"

	standard := regexp.MustCompile(pattern)
	require.True(t, standard.MatchString("aa"))
	require.True(t, standard.MatchString("aXa"))
	require.False(t, standard.MatchString("aXb"))

	matcher, err := newFastRegexMatcherWithoutCache(pattern)
	require.NoError(t, err)
	require.True(t, matcher.MatchString("aa"))
	require.True(t, matcher.MatchString("aXa"))
	require.False(t, matcher.MatchString("aXb"))
```